app/main.py

```diff
@@ -1,34 +1,126 @@
+    def get_deck(self, row: int, column: int) -> Deck:
+        for deck in self.decks:
+            if deck.row == row and deck.column == column:
+                return deck
 
-    def fire(self, row, column):
-        # Change the `is_alive` status of the deck
-        # And update the `is_drowned` value if it's needed
-        pass
+    def fire(self, row: int, column: int) -> None:
+        deck = self.get_deck(row, column)
+        deck.is_alive = False
+
+        hit_decks = 0
+        for deck in self.decks:
+            if deck.is_alive is False:
+                hit_decks += 1
+        if hit_decks == len(self.decks):
+            self.is_drowned = True
 
 
 class Battleship:
-    def __init__(self, ships):
-        # Create a dict `self.field`.
-        # Its keys are tuples - the coordinates of the non-empty cells,
-        # A value for each cell is a reference to the ship
-        # which is located in it
-        pass
-
-    def fire(self, location: tuple):
-        # This function should check whether the location
-        # is a key in the `self.field`
-        # If it is, then it should check if this cell is the last alive
-        # in the ship or not.
-        pass
+    def __init__(self, ships: list) -> None:
+        self._validate_field(ships)
+
+        field = {}
+        for x_coord in range(10):
+            for y_coord in range(10):
+                field[(x_coord, y_coord)] = None
+
+        for item in ships:
+            ship = Ship(item[0], item[1])
+            for deck in ship.decks:
+                field[(deck.row, deck.column)] = ship
+
+        self.field = field
+
+    @staticmethod
+    def _validate_field(ships: list) -> None:
+
+        if len(ships) != 10:
+            raise ValueError("There must be 10 ships.")
+
+        for ship in ships:
+            other_ships = [
+                end for temp in ships if temp != ship for end in temp
+            ]
+            for end in ship:
+                neighbours = [
+                    (end[0] - 1, end[1] - 1),
+                    (end[0] - 1, end[1]),
+                    (end[0] - 1, end[1] + 1),
+                    (end[0], end[1] - 1),
+                    (end[0], end[1] + 1),
+                    (end[0] + 1, end[1] - 1),
+                    (end[0] + 1, end[1]),
+                    (end[0] + 1, end[1] + 1),
+                ]
+
+                for neighbour in neighbours:
+                    if neighbour in other_ships:
+                        raise ValueError("The ships must not be adjacent.")
+
+        deck_size = []
+        for ship in ships:
+            deck_size.append(
+                (ship[1][0] - ship[0][0]) + (ship[1][1] - ship[0][1]) + 1
+            )
+        if sorted(deck_size) != [1, 1, 1, 1, 2, 2, 2, 3, 3, 4]:
+            raise ValueError("Wrong ship size configuration.")
+
+    def fire(self, location: tuple) -> str:
+        row = location[0]
+        column = location[1]
+        cell = self.field[location]
```

Comment:
Accessing `self.field[location]` without checking if `location` is a valid key could lead to a `KeyError`. Consider adding a check to ensure `location` is within the bounds of the field.